### ISSUE TITLE
Framework: Implement way to save images locally from any image url in the sheet

### ISSUE THREAD

**Issue by stacksavings** [StackSavingsTeam/StackSavings-Layouts#42]
The layouts app is designed to be used by non-technical users and they may use image url's from any URL, so it won't necessarily be an image that is uploaded to our FTP by the user. This presents a problem, however, as the image they use may be uploaded somewhere that does not use a CDN.

Because of this, we want a solution where any image that is in the sheet is copied locally to our project and the actual image url that the page uses to show the image is from this local copy. This will allow our CDN to then cache it along with the other content.

It would be best to have this only have to run one time per image.

**Comment by cyberdypro** [comment]
@stacksavings I have an idea about we will check current site domain with image domain, if the image doesn't store in the same domain with current site we will download that image to current server and store into the cache file (json should be better i think). So after get all values from google sheet, we will lookup in cache file with image key for get the file that downloaded & stored in local, if not exists we will download and write into the cache file with image key. how do you think ?

**Comment by stacksavings** [comment]
@cyberdypro yes I think this is good solution, I think one thing that we should add is a way to clear this cache, so it reloads the values, like with param, we want it to be secret so should be some key that is set in the properties file, would be like:

?clearimagecache=239fj39t344jgj344t49g

so "239fj39t344jgj344t49g" is like some secret key (just as example) that is set in the properties file, so if user calls with this it will clear the image cache.

**Comment by stacksavings** [comment]
@cyberdypro can you give idea as far as estimate for time on this task also?

**Comment by cyberdypro** [comment]
@stacksavings I think it may need about 8 hour

**Comment by stacksavings** [comment]
Hold off on this for now, I will have to research more.

**Comment by stacksavings** [comment]
@cyberdypro This is higher than expected as far as the estimate. Please clarify what will take this long and break this down further so that I can review before we can start.

**Comment by cyberdypro** [comment]
I have reduced the estimation, it may contains:
- Create function for make the json cache file
- Add logic for detect the domain of image item
- Add logic for download image to our server
- Add logic for clear cache
- Add logic for read image url from cache intead of sheet
So I think it may need about 4 hrs. Please advise, thanks !

**Comment by stacksavings** [comment]
@cyberdypro  ok thanks, we will do this at some point but I have some new tasks coming up that I think are more important relating to a new layout so let's keep this one on hold for right now and I will update regarding the other tasks soon.
